src/program.hpp

```diff
@@ -142,6 +142,35 @@ enum class spec_constant
     subgroup_max_size,
 };
 
+struct user_spec_constant_data {
+    std::string type;
+    uint32_t size;
+    bool set;
+    union {
+        uint8_t i8;
+        uint16_t i16;
+        uint32_t i32;
+        uint64_t i64;
+    } data;
+
+    user_spec_constant_data(std::string type, uint32_t size)
+        : type(type), size(size), set(false) {}
+    user_spec_constant_data() : type(""), size(0), set(false) {}
+
+    void init_data(size_t size, const void* value) {
+        if (type == "i8" || type == "i1") {
+            std::memcpy(&data.i8, value, size);
+        } else if (type == "i16" || type == "f16") {
+            std::memcpy(&data.i16, value, size);
+        } else if (type == "i32" || type == "f32") {
+            std::memcpy(&data.i32, value, size);
+        } else if (type == "i64" || type == "f64") {
+            std::memcpy(&data.i64, value, size);
+        }
```

Comment:
We should probably have an else case to catch errors here or at least an assert.